src/dune/file_selector.ml

```diff
@@ -1,31 +1,66 @@
 open Stdune
 
+module Predicate = struct
+  type t =
+    | Predicate of string Predicate.t
+    | Glob of Glob.t
+
+  let test = function
+    | Predicate predicate -> Predicate.test predicate
+    | Glob glob -> Glob.test glob
+
+  let compare x y =
+    match (x, y) with
+    | Predicate x, Predicate y -> Predicate.compare x y
+    | Predicate _, _ -> Lt
+    | _, Predicate _ -> Gt
+    | Glob x, Glob y -> Glob.compare x y
+
+  let hash = function
+    | Predicate predicate -> Hashtbl.hash (0, Predicate.hash predicate)
+    | Glob glob -> Hashtbl.hash (1, Glob.hash glob)
+
+  let to_dyn =
+    let open Dyn in
+    function
+    | Predicate predicate ->
+      Variant ("predicate", [ Predicate.to_dyn predicate ])
+    | Glob glob -> Variant ("glob", [ Glob.to_dyn glob ])
+
+  let encode =
+    let open Dune_lang.Encoder in
+    function
+    | Predicate predicate ->
+      record [ ("predicate", Predicate.encode predicate) ]
```

Comment:
`record` results in an extra layer of parentheses. Strangely, `variant` function doesn't exist and instead in `dep.ml` we're using `pair`.
Maybe you can add `val variant : string -> sexp list -> sexp` and use that?